### ISSUE TITLE
Cuda 9, CuDNN 7, TensorFlow 1.5 seem to work fine. any reason not to use the latest versions?

### ISSUE THREAD

**Issue by IUsedToBeAPygmy** [deepfakes/faceswap#161]
I'm working on Windows 10.
I've installed Cuda 9, CuDNN 7 and TensorFlow 1.5, and everything has been working fine for a week now.
I think it's a bit faster, but I have not measured it.
Is there any specific reason to linger with Cuda 8 and TF 1.4?

**Comment by gdunstone** [comment]
I think we should use something like `tensorflow-gpu>=1.4.0`

**Comment by facepainter** [comment]
@IUsedToBeAPygmy 

I too have a set up with CUDA 9, CuDNN 7 and TF 1.5 using Python 3.6 on Win10 64bit and it works great...execpt I can't build dlib with CUDA support against version 9 :( 

Have you had joy with this? This is all that is keeping me on v8/6 and TF 1.4...

**Comment by LordVulkan** [comment]
I have been using CUDA 9.1, cuDNN 7 and a custom build from TensorFlow master branch without trouble.

**Comment by Clorr** [comment]
I have a commit ready for that:

https://github.com/deepfakes/faceswap/commit/c1ce49ec6277ff20b8e85c30ff3a7fa3d2a90e4a

You can merge it if it's ok for you

**Comment by shadowfolder** [comment]
@facepainter i was unable install dlib via pip too. But i sucessfully installed dlib via github https://github.com/davisking/dlib via python setup.py install

**Comment by dr-benway** [comment]
works great on > CUDA9 CuDNN7 TF 1.6.0rc0 Python 3.6.2  Win10 x64

**Comment by gdunstone** [comment]
Not everyone runs Archlinux.

**Comment by dr-benway** [comment]
works great on > CUDA9 CuDNN7 TF 1.6.0rc0 Python 3.6.2  Win10 x64
training with GAN128 (increased kernel sizes to 4 in order for them to be divisible by the strides)

**Comment by dr-benway** [comment]
@facepainter @shadowfolder 
check this out for a compiled wheel > http://www.xavierdupre.fr/app/pymyinstall/helpsphinx/blog/2017/2017-07-10_dlib.html

**Comment by facepainter** [comment]
@dr-benway @shadowfolder - hmm yeah I can download/install and compile it fine...just not with CUDA support.

**Comment by LordVulkan** [comment]
> facepainter commented 18 minutes ago
> @dr-benway @shadowfolder - hmm yeah I can download/install and compile it fine...just not with CUDA support.

Can you paste the build log for dlib? I built it successfully against CUDA 9.1

**Comment by facepainter** [comment]
@LordVulkan  - Agh sorry didn't save - will try again though and post back...if you say it can be done that is good :) - kept getting errors when building the CUDA test app. From what I'd read people were saying it was not compatible so I gave up....Just to be 100% we are talking about building dlib with CUDA support on Win64 using CUDA 9.1 and cuDNN 7 for Visual Studio 2017?

**Comment by LordVulkan** [comment]
> facepainter commented 7 minutes ago •  edited 
> @LordVulkan - Agh sorry didn't save - will try again though and post back...if you say it can be done that is good :) - kept getting errors when building the CUDA test app. From what I'd read people were saying it was not compatible so I gave up....Just to be 100% we are talking about building dlib with CUDA support on Win64 using CUDA 9.1 and cuDNN 7 for Visual Studio 2017? i.e.
> 
> cmake -G "Visual Studio 15 Win64" -T host=x64 -DUSE_AVX_INSTRUCTIONS=1 -DDLIB_USE_CUDA=1

`python setup.py bdist_wheel --yes USE_AVX_INSTRUCTIONS`

If I remember correctly, it automatically detects CUDA installation and I just had to tell cmake to use gcc-6 via CC and CXX variables.

**Comment by iperov** [comment]
@IUsedToBeAPygmy can you post comparison benchmarks 
**Cuda 8, CuDNN 6 and TensorFlow 1.4** 
vs
**Cuda 9, CuDNN 7 and TensorFlow 1.5**
?

**Comment by Clorr** [comment]
@dr-benway Are you sure about the dlib version you specify? I don't remember why, but I found at some point that you have to use dlib > 19.7 with the project. (And there is wheel for it with Python 3.6)

**Comment by Clorr** [comment]
Added PR #183 please review ASAP so that I can merge...

**Comment by dr-benway** [comment]
@Clorr it certainly does work on the CPU /python3.6

**Comment by k3erg** [comment]
Just for reference:
macOS High Sierra, custom built tensorflow 1.4 with GPU support, built against CUDA 9.1 and CuDNN works ok.
Going to take a shot at tensorfow 1.5 with GPU support later today.

**Comment by Apollo122** [comment]
win10, I've installed Cuda 9, CuDNN 7 and TensorFlow 1.5. compiled dlib with cuda 9 and used faceswap without a problem.
Then i thought i should get rid of cuda 8, uninstalled it but after it couldnt import dlib anymore. said couldnt find dll or something. installed cuda 8 again and it started work again. i read that they could co-exist together, so far no problems.

**Comment by kylefrost** [comment]
@k3erg If you get 1.5 working with GPU on High Sierra, any chance you could type up a quick gist on how? I keep erroring after a few minutes, and can't figure out why.

**Comment by subzerofun** [comment]
@kylefrost  In case you haven't solved this yourself – this is a great tutorial:  [Tensorflow 1.4 Mac OS High Sierra 10.13 GPU Support](https://gist.github.com/smitshilu/53cf9ff0fd6cdb64cca69a7e2827ed0f) 

Maybe it works for TF 1.5 too. Compiling TF on macOS since Google dropped official support can be quite time consuming ... Especially if you want to have the newest CUDA + cudnn versions. After a lot of wasted hours trying to get TF 1.4, the newest PyTorch and Torch all working with CUDA 9.1 i thought i'd share my setup plus a little hack to switch between CUDA versions.

**This is not meant to just fix problems with Tensorflow – it can come in handy if some ML frameworks work with CUDA version X, but not version Y etc.**

Download all CUDA (full DMG) installers, copy them somewhere safe if you later need to reinstall a specific version. Do the same with all CUDNN versions. Install CUDA 8.0, copy the according **cudnn** files into` /usr/local/cuda` (include and lib, like the documentation explains). **Hint**: The folder in `/usr/local/cuda/lib` is not symlinked to `/Developer/NVIDIA/CUDA-X.X/lib` – but most **files in it** are symlinked. I always copy the cudnn lib files also to the NVIDIA lib folder.

Rename the /usr/local/cuda folder to cuda80. Install CUDA 9.0, copy according cudnn files. Make sure the cudnn version is compatible with CUDA 9.0.

Rename /usr/local/cuda to cuda90. Finally install CUDA 9.1 + copy matching cudnn. The newest version will therefore point to /usr/local/cuda (don't rename to cuda91!).

Clean up your bash source file: Make sure that all environment variables point to /usr/local/cuda lib or include – correct everything with /Developer/NVIDIA/CUDA-X.X to /usr/local/cuda.

That way you can always revert back to an older CUDA + cudnn version: You just have to rename the /usr/local/cudaXX folder back to ../cuda. Only downside: When switching versions you (sometimes) need to reinstall the correct CUDA drivers/kexts (from the install dmg, only select drivers – not the whole toolkit).

Sounds a little complicated, but the advantage is that you can write a bash function that changes all env variables from /usr/local/cuda to /usr/local/cudaXX when needed. You can then reset your terminal session and try to install TF/PyTorch/etc. with another CUDA version.

I have now got CUDA 9.1 with TF 1.4 working on macOS 10.12.5. PyTorch with GPU support needed CUDA 8.0, that's where the bash function came in handy. The regular "old" Torch also fails to compile with CUDA 9.0 or 9.1, so the trick with the env variables also helped me successfully install it with CUDA 8.0. As long as you set the shell variables correctly, multiple different CUDA installs are possible.

--- 

Coming back to you initial question: It's always a good idea to check if the `LD_LIBRARY_PATH` and `DYLD_LIBRARY_PATH` vars don't point to older installs in /Developer/NVIDIA/... (sort out the bash source file in general). Best to set entries like `CUDA_ROOT`, `CUDA_DIR`, `CUDA_TOOLKIT_PATH`, etc. to `/usr/local/cuda/`(lib, include, etc) because then all files are symlinked to the current install.
